text/0000-safety-tags.md

```diff
@@ -0,0 +1,485 @@
+for unsafe APIs. While these comments are generally human-readable, they can be ambiguous and
+laborious to write. Even the current best practices in the Rust standard library are somewhat ad hoc
+and informal. Moreover, safety comments are often repetitive and may be perceived as less important
+than the code itself, which makes them error-prone and increases the risk that reviewers may
+overlook inaccuracies or missing safety requirements.
+
+For instance, a severe problem may arise if the safety requirements of an API change over time:
+downstream users may be unaware of such changes and thus be exposed to security risks. Therefore, we
+propose to improve the current practice of writing safety comments by making them checkable through
+a system of safety tags. These tags are designed to be:
+
+* Compatible with existing safety documentation: Safety tags should be expressive enough to
+  represent current safety comments, especially as rendered in today's rustdoc HTML pages.
+* Usable by compiler tools for safety checking: If no safety tags are provided for an unsafe API,
+  lints should be emitted to remind developers to provide safety requirements. If a safety tag is
+  declared for an unsafe API but not discharged at a callsite, lints should be emitted to alert
+  developers about potentially overlooked safety requirements.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Syntax of Safety Tags
+
+Syntax of a safety tag is defined as follows:
+
+```text
+SafetyTag -> `#` `[` `clippy::safety` `{` Tags `}` `]`
+
+Tags -> Tag (`;` Tag)*
+
+Tag -> ID (`,` ID)* (`:` LiteralString)?
+
+ID -> SingleIdent | SimplePath
+```
+
+Here are some tag examples:
+
+```rust
+#[clippy::safety { SP }]
+#[clippy::safety { SP1, SP2 }]
+
+#[clippy::safety { SP1: "reason" }]
+#[clippy::safety { SP1: "reason"; SP2: "reason" }]
+
+#[clippy::safety { SP1, SP2: "shared reason for the two SPs" }]
+#[clippy::safety { SP1, SP2: "shared reason for the two SPs"; SP3 }]
+#[clippy::safety { SP3; SP1, SP2: "shared reason for the two SPs" }]
+```
+
+`#[clippy::safety]` is a [tool attribute] that you attach to an unsafe function (or to an expression
+that performs unsafe calls). Take [`ptr::read`]: its safety comment lists three requirements, so we
+create three corresponding tags on the function declaration and mark each one off at the call site.
+
+```rust
+#[clippy::safety { ValidPtr, Aligned, Initialized }] // defsite
+pub unsafe fn read<T>(ptr: *const T) -> T { ... }
+
+#[clippy::safety { ValidPtr, Aligned, Initialized }] // callsite
+unsafe { read(ptr) };
+```
+
+We can also attach comments for a tag or a group of tags to clarify how safety requirements are met:
+
+```rust
+#[clippy::safety {
+  ValidPtr, Aligned, Initialized: "addr range p..p+n is properly initialized from aligned memory"
+  InBounded, ValidNum: "`n` won't exceed isize::MAX here, so `p.add(n)` is fine";
+}]
+for _ in 0..n {
+    unsafe {
```

Comment:
The `#[clippy::safety]` attached to the `for` expression rather than the inner `unsafe`, is this expected? i'm not sure if allowing `#[clippy::safety]` be applied "at a distance" is a good idea

Reply:
Yes. It's the expected behavior, as it's meaningful to explain how invariants are upheld in loops. [Previous discussion.](https://rust-lang.zulipchat.com/#narrow/channel/136281-t-opsem/topic/Safety.20Property.20System/near/530636749)